HIP-1.md

```diff
@@ -0,0 +1,146 @@
+        bytes calldata _payload
+    ) external;
+}
+
+interface IInterchainSecurityModule {
+    // Used by relayers to determine which off-chain data to inject, and
+    // how to format it.
+    // The first byte must contain the module "type", which informs the relayer
+    // how to interpret the remaining bytes.
+    function moduleData() external view returns (bytes memory);
+
+    // Called by the Inbox to determine whether the provided root
+    // is valid to verify proofs against.
+    function accept(
+        bytes32 _root,
+        bytes calldata _message,
+        bytes calldata _data
+    ) external returns (bool);
+}
+
+interface IInbox {
+
+    // Called by relayers to deliver messages to the recipient.
+    // Verifies the merkle proof of `_message` against `_root` and
+    // calls `recipient.interchainSecurityModule().accept()` before
+    // calling `recipient.handle()`.
+    function process(
+        bytes32 _root,
+        bytes32[32] calldata _proof,
+        bytes calldata _message,
+        bytes calldata _data
+    ) external;
+}
+```
+
+### **Rationale**
+
+Sovereign consensus allows for a modular approach to interchain security.
+This allows the core protocol to be forward compatible, ensuring that Hyperlane can support new security models as they are developed.
+Furthermore, it allows applications to select, configure, or invent security models that offer the most appropriate tradeoffs.
+
+There were many ways to build this feature.
+We discuss a few of the design decisions made below.
+
+#### Location of merkle proof verification
+
+The ability for relayers to inject arbitrary data into the `ISM` presents a natural opportunity to move the merkle proof out of the `Inbox.process()` interface and into the `ISM`.
+In this model, the `Inbox` would be responsible only for replay protection.
+
+```
+interface IInterchainSecurityModule {
+    function moduleData() external view returns (bytes memory);
+
+    function accept(
+        bytes32 _root,
+        bytes32[32] calldata _proof,
+        bytes calldata _message,
+        bytes calldata _data
+    ) external returns (bool);
+}
+
+interface IInbox {
+    function process(bytes calldata _message) external;
+}
+```
+
+This approach has the advantage of being even more modular than the proposed solution, as the `Inbox` is unopinionated about how messages are sent.
+This allows applications to create `ISMs` that do not incorporate the incremental merkle tree.
+As a simple example, one could build an `ISM` that is a multisig on individual messages, as opposed to a merkle root.
+
+The drawback is that those `ISMs` which wish to incorporate the incremental merkle tree into their security model must do that explicitly.
+We imagine most applications to prefer the censorship resistance and economies of scale that using the incremental merkle tree allows for.
+Thus, we propose leaving proof verification in the `Inbox` so that all `ISM` developers get this "for free".
```

Comment:
It's not that hard if you know what you're doing, but perhaps the hard part is knowing what you're doing?

I agree it's more modular.

For example, you could imagine a scheme in which validators (regular or optimistic) attest to individual messages. If you want to continue to have fraud proofs and slashing, you can go about it in two ways:
1. Have the Outbox store `sent[messageId] = true`
2. Use the incremental merkle tree. So long as the message includes the leaf index (which ofc the protocol can mandate), you can prove fraud by providing a merkle proof that `leaf[leaf_index] != messageId`.

The ISM for both of these approaches looks exactly the same, the difference is only in the Outbox on the origin chain, for which (1) appears to be ~10-20k gas cheaper.

You could use the same approaches, I think, for zkp based approaches.

The hard part, I suppose, is that you'd need the sender to specify which storage modality they'd like to send with. Validators/relayers would need to support multiple storage modalities.

I'm not sure exactly how to make that "seamless" for the sender...

Reply:
@nambrot there is something attractive about putting the merkle proof verification in the ISM.

At present, if you want to be able to slash for censorship, you need to emit the merkle proof in an event, so that you can use the merkle proof as evidence that some prior leaf in the tree had been modified.

The logical place to emit this event is in the same place as the code that verifies the merkle proof. And emitting the event is expensive. So if you put the merkle proof verification in the mailbox contract, you're stuck with that cost even if you want to use some security model for which slashing for censorship is not really relevant (e.g. zkp based light clients).


That said, this is an argument for putting verification of the merkle proof in the ISM, and not an argument for modularizing the storage modality of the Outbox, which I think is still worth discussing.

Reply:
I'm not sure I quite followed, but yeah just on the basis that zkp wouldn't really use the merkle tree I would move it into the ISM so that a zkp-ISM does not incur that overhead unnecessarily

Reply:
Put differently, the benefit of censorship resistance and fraud proofs makes sense for a subset of all security models and therefore ISMs and thus should be encapsulated there

Reply:
@nambrot the problem is that ISMs are a destination chain concept and now you're talking about requiring knowledge of the destination chain ISM on the origin chain. We don't really have a precedent for how to deal with this

Reply:
I think ripping out merkle trees will be way too confusing for people

Another thing I like about having merkle trees in the core protocol is that changing ISMs by default would still allow old messages to be processed. E.g. if an app is using ISM A, some messages are sent but not yet processed at their destination, and the app switches to ISM B, then ISM B doesn't need relayers to look back at any unprocessed messages that may need to be re-signed according to the rules of ISM B-- ISM B just needs to sign the latest merkle root and those old unprocessed messages can be processed as normal

Reply:
> Another thing I like about having merkle trees in the core protocol is that changing ISMs by default would still allow old messages to be processed

@tkporter I'm not sure that what you're saying is any different if the merkle proof verification is in the ISM vs the `Mailbox`. Either way the relayer will need to call `Mailbox.process()` with:
- A signed checkpoint
- The message
- A merkle proof of the message against the signed checkpoint

Reply:
Or maybe I misinterpreted what you were objecting to. IMO moving the merkle tree into the ISM is a decent compromise

Reply:
I meant in a world where someone's using an ISM that doesn't use Merkle trees as an accumulator for messages (e.g. imagine just signing individual messages), if they change ISMs then any messages that the old ISM had signed that weren't processed now need to be signed by the new ISM

but I think I'm not understanding what the current thinking is around merkle trees with proofs being in ISMs / if the imagined situation where an ISM doesn't sign an accumulator is something worth caring about. Need to read the v2 code